static/app/views/insights/database/utils/formatMongoDBQuery.tsx

```diff
@@ -0,0 +1,145 @@
+ */
+export function formatMongoDBQuery(query: string, command: string) {
+  const queryObject = JSON.parse(query);
+
+  const tokens: ReactElement[] = [];
+
+  insertToken(tokens, '{ ');
+
+  // Start by finding the key-value pair that represents the operation on the collection
+  // and insert it as the first token
+  const commandKey = Object.keys(queryObject).find(
+    key => key.toLowerCase() === command.toLowerCase()
+  );
+
+  // If the command was somehow not found in the query, return it as a raw string since no formatting can be applied
+  if (!commandKey) {
+    return query;
+  }
+
+  const collection = queryObject[commandKey];
+
+  processAndInsertKeyValueTokens(tokens, commandKey, collection, true);
+
+  // Create the remaining tokens by iterating over all keys and skipping the command key
+  Object.keys(queryObject).forEach(key => {
+    if (key.toLowerCase() !== command.toLowerCase()) {
+      const value = queryObject[key];
+      processAndInsertKeyValueTokens(tokens, key, value);
+    }
+  });
+
+  // Pop the last token since it will be an extra comma
+  tokens.pop();
+  insertToken(tokens, ' }');
+
+  return tokens;
+}
+
+type JSONValue = string | number | object | boolean | null;
+
+function processAndInsertKeyValueTokens(
+  tokens: ReactElement[],
+  key: string,
+  value: JSONValue | JSONValue[],
+  // Use isBold only for non-object kv pairs
+  isBold?: boolean
+) {
+  // Wrapper function for better readability
+  const _insertToken = (token: string) => insertToken(tokens, token, isBold);
+  // Case 1: Value is null
+  if (!value) {
+    _insertToken(`"${key}": null`);
+    _insertToken(', ');
+    return;
+  }
+
+  // Case 2: Value is a string
+  if (typeof value === 'string') {
+    _insertToken(`"${key}": "${value}"`);
+    _insertToken(`, `);
+    return;
+  }
+
+  // Case 3: Value is one of the other primitive types
+  if (typeof value === 'number' || typeof value === 'boolean') {
+    _insertToken(`"${key}": ${value}`);
+    _insertToken(`, `);
+    return;
+  }
+
+  // Case 3: Value is an array
+  if (Array.isArray(value)) {
+    _insertToken(`"${key}": [`);
+
+    value.forEach(item => {
+      if (!item) {
+        _insertToken('null');
+        _insertToken(', ');
+      } else if (
+        typeof item === 'string' ||
+        typeof item === 'number' ||
+        typeof item === 'boolean'
+      ) {
+        _insertToken(`${item}`);
+        _insertToken(', ');
+      } else {
+        // We must recurse if item is an object
+        if (Object.keys(item).length === 0) {
+          processAndInsertKeyValueTokens(tokens, key, '{}');
+        } else {
+          _insertToken('{ ');
+          Object.keys(item).forEach(_key => {
+            processAndInsertKeyValueTokens(tokens, _key, item[_key]);
+          });
+          // Pop the trailing comma
+          tokens.pop();
+
+          _insertToken(' }');
+          _insertToken(', ');
+        }
+      }
```

Comment:
There's a fair amount of repetition here with the above that makes me think that this could be structured differently. More importantly I think, it doesn't handle cases like arrays inside arrays.

If you had a function that converts a `JSONValue` to a `ReactElement[]` (handling each type in the `JSONValue` union, recursing as necessary), I think you could plug that into `formatMongoDBQuery` a little easier and cut down on the repetition (and handle the missing case).

I'd be okay saving that for a followup PR, but if so, I'd still add a case to this one before merging to cover the array-in-array case, at least to bail out somehow. Currently they'll be treated as objects, which will break.